File: modules/dotted-underline-text/ios/DottedUnderlineTextModule.swift
import ExpoModulesCore
import UIKit

/**
 * JS-side representation of `UnderlineRange`. Expo's `Record` machinery
 * decodes plain JS objects into this when used as a Prop type.
 */
struct UnderlineRangeRecord: Record {
  @Field var start: Int = 0
  @Field var end: Int = 0
  @Field var style: String?
  @Field var color: UIColor?
  @Field var thickness: Double?
}

public class DottedUnderlineTextModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DottedUnderlineText")

    View(DottedUnderlineTextView.self) {
      Events("onPress", "onRangeTap")

      Prop("text") { (view: DottedUnderlineTextView, value: String?) in
        view.text = value ?? ""
      }
      Prop("fontSize") { (view: DottedUnderlineTextView, value: Double?) in
        if let v = value { view.fontSize = CGFloat(v) }
      }
      Prop("color") { (view: DottedUnderlineTextView, value: UIColor?) in
        if let v = value { view.textColor = v }
      }
      Prop("fontFamily") { (view: DottedUnderlineTextView, value: String?) in
        view.fontFamily = value
      }
      Prop("fontWeight") { (view: DottedUnderlineTextView, value: String?) in
        view.fontWeight = value
      }
      Prop("letterSpacing") { (view: DottedUnderlineTextView, value: Double?) in
        view.letterSpacing = CGFloat(value ?? 0)
      }
      Prop("lineHeight") { (view: DottedUnderlineTextView, value: Double?) in
        view.lineHeight = CGFloat(value ?? 0)
      }
      Prop("textAlign") { (view: DottedUnderlineTextView, value: String?) in
        switch value {
        case "center": view.textAlign = .center
        case "right":  view.textAlign = .right
        case "left":   view.textAlign = .left
        case "justify": view.textAlign = .justified
        default:       view.textAlign = .natural
        }
      }
      Prop("underlineColor") { (view: DottedUnderlineTextView, value: UIColor?) in
        view.underlineColor = value
      }
      Prop("underlineStyle") { (view: DottedUnderlineTextView, value: String?) in
        view.underlineStyle = value ?? "dotted"
      }
      Prop("underlineThickness") { (view: DottedUnderlineTextView, value: Double?) in
        view.underlineThickness = CGFloat(value ?? 1)
      }
      Prop("ranges") { (view: DottedUnderlineTextView, value: [UnderlineRangeRecord]?) in
        guard let value = value else {
          view.ranges = nil
          return
        }
        view.ranges = value.map { r in
          UnderlineRangeSpec(
            start: r.start,
            end: r.end,
            style: r.style ?? "dotted",
            color: r.color
          )
        }
      }
      Prop("selectable") { (view: DottedUnderlineTextView, value: Bool?) in
        view.selectable = value ?? true
      }
      Prop("accessibilityLabel") { (view: DottedUnderlineTextView, value: String?) in
        view.accessibilityLabel = value
      }
    }
  }
}

File: modules/dotted-underline-text/android/src/main/java/expo/modules/dottedunderlinetext/DottedUnderlineTextModule.kt
package expo.modules.dottedunderlinetext

import android.graphics.Color
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

/**
 * Range record forwarded from JS via the `ranges` prop. The Expo Records
 * machinery deserializes plain JS objects into this type automatically when
 * referenced as a Prop value type.
 */
class RangeRecord : Record {
  @Field var start: Int = 0
  @Field var end: Int = 0
  @Field var style: String? = null
  @Field var color: String? = null
  @Field var thickness: Double? = null
}

class DottedUnderlineTextModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DottedUnderlineText")

    View(DottedUnderlineTextView::class) {
      Events("onPress", "onRangeTap")

      Prop("text") { view: DottedUnderlineTextView, value: String? ->
        view.setText(value ?: "")
      }
      Prop("fontSize") { view: DottedUnderlineTextView, value: Double? ->
        if (value != null) view.setFontSize(value.toFloat())
      }
      Prop("color") { view: DottedUnderlineTextView, value: Int? ->
        if (value != null) view.setTextColor(value)
      }
      Prop("fontFamily") { view: DottedUnderlineTextView, value: String? ->
        view.setFontFamily(value)
      }
      Prop("fontWeight") { view: DottedUnderlineTextView, value: String? ->
        view.setFontWeight(value)
      }
      Prop("letterSpacing") { view: DottedUnderlineTextView, value: Double? ->
        if (value != null) view.setLetterSpacing(value.toFloat())
      }
      Prop("lineHeight") { view: DottedUnderlineTextView, value: Double? ->
        if (value != null) view.setLineHeightPx(value.toFloat())
      }
      Prop("textAlign") { view: DottedUnderlineTextView, value: String? ->
        view.setTextAlign(value)
      }
      Prop("underlineColor") { view: DottedUnderlineTextView, value: Int? ->
        if (value != null) view.setUnderlineColor(value)
      }
      Prop("underlineStyle") { view: DottedUnderlineTextView, value: String? ->
        view.setUnderlineStyle(value ?: "dotted")
      }
      Prop("underlineThickness") { view: DottedUnderlineTextView, value: Double? ->
        view.setUnderlineThickness((value ?: 1.0).toFloat())
      }
      Prop("ranges") { view: DottedUnderlineTextView, value: List<RangeRecord>? ->
        if (value == null) {
          view.setRanges(null)
        } else {
          val descriptors = value.map { r ->
            val parsedColor = try {
              if (!r.color.isNullOrEmpty()) Color.parseColor(r.color) else Color.BLACK
            } catch (_: IllegalArgumentException) {
              Color.BLACK
            }
            UnderlineDescriptor(
              start = r.start,
              end = r.end,
              color = parsedColor,
              thicknessDp = (r.thickness ?: 1.0).toFloat(),
              isDotted = (r.style ?: "dotted") == "dotted"
            )
          }
          view.setRanges(descriptors)
        }
      }
      Prop("selectable") { view: DottedUnderlineTextView, value: Boolean? ->
        view.setSelectableText(value ?: true)
      }
      Prop("accessibilityLabel") { view: DottedUnderlineTextView, value: String? ->
        view.setAccessibilityLabelText(value)
      }
    }
  }
}
